### PR TITLE
feat(torch): clip gradient options for all optimizers

### PR DESCRIPTION
**Pull request by fantes**
before this PR, only RANGER was clipping gradient if asked for, we now allow for clipping gradients (both to values and to norm) for all torch optimizers

### CODE REVIEW

File: docs/api.md [hunk 1]
```diff
@@ -690,8 +690,9 @@ iterations           | int          | yes      | N/A     | Max number of solver'
 snapshot             | int          | yes      | N/A     | Iterations between model snapshots
 snapshot_prefix      | string       | yes      | empty   | Prefix to snapshot file, supports repository
 solver_type          | string       | yes      | SGD     | from "SGD", "ADAGRAD", "NESTEROV", "RMSPROP", "ADADELTA", "ADAM",  "AMSGRAD", "RANGER", "RANGER_PLUS", "ADAMW", "SGDW", "AMSGRADW" (*W version for decoupled weight decay, RANGER_PLUS is ranger + adabelief + centralized_gradient)
-clip                 | bool         | yes      | false (true if RANGER* selected) | clip gradients, implemented only in ranger
-clip_value           | real         | yes      | 5.0     | value for  clipping gradients (used only by RANGER)
+clip                 | bool         | yes      | false (true if RANGER* selected) | clip gradients
+clip_value           | real         | yes      | 5.0     | value for  clipping gradients
```
Comment (BynaryCobweb) [inline docs/api.md:694]: `gradients with absolute value greater than clip_value will be clipped` is more explicit and better imo

Same for `clip_norm`

File: tests/ut-torchapi.cc [hunk 2]
```diff
@@ -987,3 +987,74 @@ TEST(torchapi, service_train_ranger)
...
+  ASSERT_EQ(created_str, joutstr);
+
+  // Train
+  std::string jtrainstr
+      = "{\"service\":\"imgserv\",\"async\":false,\"parameters\":{"
+        "\"mllib\":{\"solver\":{\"iterations\":"
+        + iterations_resnet50
+        + ",\"base_lr\":1e-5,\"iter_size\":4,\"solver_type\":\"RMSPROP\","
+          "\"clip\":true,\"test_interval\":"
+        + iterations_resnet50
```
Comment (BynaryCobweb) [inline tests/ut-torchapi.cc:1013]: Maybe we don't need as much iterations to test clipping? Could we reduce the number of iterations and just check that the loss is lower ?

File: src/backends/torch/torchsolver.cc [hunk 3]
```diff
@@ -136,9 +133,38 @@ namespace dd
...
+        if (_clip_value >= 0.0)
+          this->_logger->info("clip_value: {}", _clip_value);
+        if (_clip_norm >= 0.0)
+          this->_logger->info("clip_norm: {}", _clip_value);
+        if (_clip_norm < 0.0 || _clip_value < 0.0)
+          {
+            this->_logger->warn(
+                "Gradient clipping selected, but no value given, "
+                "will clip gradient norm to {}",
+                DEFAULT_CLIP_NORM);
```
Comment (BynaryCobweb) [inline src/backends/torch/torchsolver.cc:151]: Same thing for DEFAULT_CLIP_VALUE?

File: docs/api.md (hunk 1 above)
Comment (fantes) [reply docs/api.md:694]: ok done

File: src/backends/torch/torchsolver.cc (hunk 3 above)
Comment (fantes) [reply src/backends/torch/torchsolver.cc:151]: if nothing is given, currently code only clips norm of gradient, not value

Comment (BynaryCobweb) [reply src/backends/torch/torchsolver.cc:151]: ok
but DEFAULT_CLIP_VALUE is never used in the code

File: tests/ut-torchapi.cc (hunk 2 above)
Comment (fantes) [reply tests/ut-torchapi.cc]: done somthing

File: src/backends/torch/torchsolver.cc (hunk 3 above)
Comment (fantes) [reply src/backends/torch/torchsolver.cc]: if not defined, we now clip both absolute values and norm

**Review by BynaryCobweb — APPROVED** [review]

**Review by sileht — APPROVED** [review]

**Pull request merged**